fixes and improvements to Android Blob Support

giantslogik opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

This PR fixes https://github.com/facebook/react-native/issues/48657 and https://github.com/facebook/react-native/issues/48762

- Memory usage for creating a blob lowered from _( nearest 2^N greater than the file size)+1KB_ previously To _(file size)+1KB_. This is achieved by avoiding internal calls to java.io.ByteArrayOutputStream.ensureCapacity by creating a buffer based on file size. 
- Increases the max file size a Blob can be  created from before hitting a OutOfMemoryException. 
- Major performance increase by avoiding buffer copying 
- fix creating Blobs from Android 'content://' scheme urls was failing on the js side due to [this check](https://github.com/JakeChampion/fetch/blob/ba5cf1ed2e02ebb96fa1e60b4fd2eb04071b60e4/fetch.js#L547)

## Changelog:

[ANDROID] [CHANGED] - Speed and Memory usage improvements to Android Blob support
[ANDROID] [FIXED] - Creating of Blobs from large files now works. File size can now be upto available (free) heap size. 
[ANDROID] [FIXED] - fix fetch of content scheme uris failing on Android.

## Test Plan:

Used the App here to test Android Blob creation. https://github.com/giantslogik/blob-large-file-fetch.

tdn120 commented:
Thanks for putting this up!  A few things:

First, I think these are 2 discrete changes, and should probably be 2 separate PRs.  I'll comment on the blob changes for now.

I can see how ByteArrayOutputStream's `grow()` logic becomes problematic in these cases, but I'm not sure this is the right solution.  I'm most concerned about the use of [InputStream.available()](https://developer.android.com/reference/java/io/InputStream#available()):
> Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking, which may be 0, or 0 when end of stream is detected. The read might be on the same thread or another thread. A single read or skip of this many bytes will not block, but may read or skip fewer bytes.
>
>Note that while some implementations of InputStream will return the total number of bytes in the stream, many will not. It is never correct to use the return value of this method to allocate a buffer intended to hold all data in this stream.

Also, the addition of `prevBuffer` to the logic also makes the code much harder to understand, I'd try to avoid it if possible (or explain why it's needed).

Please also try to reformat the code, looks like there's new indenting on some lines that otherwise wouldn't be changed.

^ This is all from a pure code perspective, there might be other ways to work around the OOM for your use case.  I'll see if the team has any suggestions.

giantslogik commented:
> First, I think these are 2 discrete changes, and should probably be 2 separate PRs. I'll comment on the blob changes for now.

 This is a single change as recent Android versions typically returns 'content' scheme urls rather than 'file' urls.
I had to jump through hoops to obtain a file url for test. (Used a fork of react-native-fs to convert url schemes, in my testing.)

> 
> I can see how ByteArrayOutputStream's `grow()` logic becomes problematic in these cases, but I'm not sure this is the right solution. I'm most concerned about the use of [InputStream.available()](https://developer.android.com/reference/java/io/InputStream#available()):
> Also, the addition of `prevBuffer` to the logic also makes the code much harder to understand, I'd try to avoid it if possible (or explain why it's needed).

 The 'prevBuffer' was specifically added to handle InputStream.available() returning 0 or less than the actual content size.

The code tries to return the buffer without writing to  ByteArrayOutputStream if InputStream.available() returns the whole File size. Otherwise it uses the ByteArrayOutputStream.  

In my testing InputStream.available() always returned the entire file size. Since this is not a hard guarantee, I have tested the code for various available() behaviour by hardcoding:

 `byte[] buffer = new byte[Math.max(1024,is.available()*N)];//where N is in the interval [0,1]`

Please let me know your thoughts, and any suggestions from the team.  I'll wait for them before i run a another pass at the PR.

giantslogik commented:
Just noting that the _Inputstream_  for (some) content urls/file urls is an _instanceof FileInputStream_.
The documentation for available is different for FileInputStream:
https://docs.oracle.com/javase/8/docs/api/java/io/FileInputStream.html#available--

> Returns an estimate of the number of remaining bytes that can be read (or skipped over) from this input stream without blocking by the next invocation of a method for this input stream. Returns 0 when the file position is beyond EOF. The next invocation might be the same thread or another thread. A single read or skip of this many bytes will not block, but may read or skip fewer bytes.
In some cases, a non-blocking read (or skip) may appear to be blocked when it is merely slow, for example when reading large files over slow networks.

The PR addresses cases where the _Inputstream_  is/isn't an _instanceof FileInputStream_.

tdn120 commented:
Ok haven't gotten many suggestions, probably best to spruce this up and then I can pull it in for internal review.

For the content:// vs. file:// issue, the current solution is a bit heavy-handed.  I would look to see why 'file' is working without the extra `200` response.  One thing that pops out to me is specific behavior for "file" in BlobModule:

[BlobModule.java#L245](https://github.com/facebook/react-native/blob/v0.77.0/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L245)
[BlobModule.java#L266](https://github.com/facebook/react-native/blob/v0.77.0/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L266)

We might need something similar for "content".  If not, and there's really no way around making this blanket change, we'll need to verify this behavior across all existing uris and handlers to ensure there are no regressions or side effects.

For the `prevBuffer` approach, I'm concerned about the deviation from conventional wisdom on network buffers.  It seems on the surface that simply addressing the `grow()` issue, without abandoning the smaller buffer altogether, would be the way to go - whether that means using different output streams / buffers or rolling our own.

I'm also not clear on the specific purpose behind using `available()` - that is seemingly used to avoid blocking, but we're trying to load the entire file here, so its utility is unclear.  And in the worst case, if `available()` returned `1`, we'd be using a 1 byte buffer, which would probably wreck performance.  It almost seems like [readAllBytes()](https://developer.android.com/reference/java/io/InputStream#readAllBytes()) might work better for your approach, but you can see the docs recommend against using it for this scenario:
> Note that this method is intended for simple cases where it is convenient to read all bytes into a byte array. It is not intended for reading input streams with large amounts of data.

Whichever way you end up, if you can spruce up the formatting and add comments about the approach and your reasoning, I can pull it in and see what the rest of the team thinks.

giantslogik commented:
@tdn120 

> For the content:// vs. file:// issue, the current solution is a bit heavy-handed. I would look to see why 'file' is working without the extra `200` response.

'file://' works without the `200` response because [whatwg-fetch has special handling hardcoded here](https://github.com/JakeChampion/fetch/blob/ba5cf1ed2e02ebb96fa1e60b4fd2eb04071b60e4/fetch.js#L547). I think 'content://' scheme handling is beyond the scope of whatwg-fetch. Not sure if this can be pushed to the dependency instead of fixing here.

>  If not, and there's really no way around making this blanket change, we'll need to verify this behavior across all existing uris and handlers to ensure there are no regressions or side effects.

   Noted. To avoid regressions and side effects , I will wrap the code in `if ("content".equals(contentUri.getScheme()))`.  

> One thing that pops out to me is specific behavior for "file" in BlobModule:
[BlobModule.java#L245](https://github.com/facebook/react-native/blob/v0.77.0/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L245)
[BlobModule.java#L266](https://github.com/facebook/react-native/blob/v0.77.0/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/blob/BlobModule.java#L266)

These 'file' specific behaviour get the last modified and filename as required for the Blob js specification for files. However android 'content' scheme hides the underlying file path and stat details for security, so these behaviours don't apply. 

> For the prevBuffer approach, I'm concerned about the deviation from conventional wisdom on network buffers.

A 'content' scheme url may be handled by an _androidx.core.content.FileProvider_ or a [ContentProvider](https://developer.android.com/reference/android/content/ContentProvider). 

In the case where the 'content://' scheme is served by _FileProvider_, the `InputStream` is an `instanceof FileInputStream`, and `available()` typically returns the entire file size.  When served by a ContentProvider , its possible the InputStream has an underlying network buffer or some other (non File / network) mechanism to serve the content. 

>I'm also not clear on the specific purpose behind using available() - that is seemingly used to avoid blocking, but we're trying to load the entire file here, so its utility is unclear. And in the worst case, if available() returned 1, we'd be using a 1 byte buffer, which would probably wreck performance.

   ` available()` is used to read the entire block of data that can be read non-blocking.  It is orders of magnitude faster than copying the bytes 1024 bytes at a time, for both files / network buffers. **If** `available()` return file size, we have the entire file in an array and can return the array, without an additional copy to a `ByteArrayOutputStream`. **Else** we fall back to copying the `InputStream` to `ByteArrayOutputStream`.

As you have noted available() may return 0,1 or some fraction of total content size.  The PR uses a minimum buffer of 1024 bytes via `new byte[Math.max(1024,is.available())]` . Worst case performance should be identical to the earlier code as it too used a 1024 byte buffer. The new code handles both file or network `InputStream`s with the same baseline behaviour/performance as before.

giantslogik commented:
@tdn120 
> prevBuffer approach

 The reason for prevBuffer / two buffers, is to be able to perform an additional _read()_ to determine if we have reached EOF (or end of stream) after the initial _read()_ without copying to`ByteArrayOutputStream`.


1.  Initial read to buffer.
2.  Swap buffer / prevBuffer
3.  Read to buffer.
4.  If EOF **AND** prevBuffer contains entire stream `return prevBuffer;`.
5.  Copy prevbuffer to `ByteArrayOutputStream`
6.  IF EOF return `byteBuffer.toByteArray();`
7.  Repeat from step 2.

giantslogik commented:
@tdn120 
I've updated the PR.

- The extra 200 response will be generated only for 'content:' scheme urls, to prevent regressions and side effects in existing code. i.e. 'file:' scheme urls.
- reformated  the code, to avoid indenting existing lines
- Added comments explaining the working of `getBytesFromUri`

tdn120 commented:
Thanks for the updates!  I'll pull this version in and see how it works.

giantslogik commented:
@tdn120 Any updates on the PR / Issue ?

tdn120 commented:
cc @javache @rshest @cortinico - maybe one of you can assist.

javache commented:
Hey! I'll look into picking this up again. Can you move the changes to NetworkingModule to a separate PR? I'm not entirely clear on the impact of those.

giantslogik commented:
> This was a single change as recent Android versions typically returns 'content' scheme urls rather than 'file' urls.

@javache This PR fixes blob support from Android urls. Modern android due to security concerns only returns 'content' urls for local resources ( not 'file' urls ). Unfortunately content urls crash the networking module. (as whatwg-fetch) does not understand them. Spliting this into two PR's would not fix the crash, and one PR would not be useful without the other.

cortinico commented:
> Spliting this into two PR's would not fix the crash, and one PR would not be useful without the other.

It would however make it easier for us to review and land separately to make sure it doesn't impact the internal infrastructure. So if you could create 2 separate PRs one on top of the other, it woudl make it easier for this whole change to be included
